Improve CupertinoPopupSurface appearance

davidhicks980 opened the pull request:
This PR makes the CupertinoPopupSurface more vibrant. Also, the gaussian kernel was switched to 30 from 20 based on  comparisons. 

@dkwingsmt - Looking forward to your dialog and fixes!


After is on bottom:
<img width="939" alt="image" src="https://github.com/flutter/flutter/assets/59215665/bd020c9b-af87-4342-9a9f-c9f8f7693456">

Notably, because the borders are very transparent, the new version looks more colorful in the sample screencap than it actually is. As such, focus on the individual colors to get a feel for the change.

| actual | old | new
|--|--|--|
| <img width="30" alt="image" src="https://github.com/flutter/flutter/assets/59215665/7de2801d-a2cc-44a4-a660-2692889fed63"> | <img width="28" alt="image" src="https://github.com/flutter/flutter/assets/59215665/48689d82-af15-4612-b4ab-75d584e9b094"> | <img width="30" alt="image" src="https://github.com/flutter/flutter/assets/59215665/7c1075ec-b815-47e0-b822-65a2b63497a0"> |
| <img width="24" alt="image" src="https://github.com/flutter/flutter/assets/59215665/2eeefe25-2e1d-4a79-b748-4925d950b9a2"> | <img width="26" alt="image" src="https://github.com/flutter/flutter/assets/59215665/68a2694c-d943-4563-9b5e-9e86e2ee1d58"> | <img width="28" alt="image" src="https://github.com/flutter/flutter/assets/59215665/1932b90a-1719-40f5-828e-41ceafd59e26"> |
<img width="22" alt="image" src="https://github.com/flutter/flutter/assets/59215665/0bd22c64-dd37-4262-a7e8-ed610151ab7a"> |  <img width="28" alt="image" src="https://github.com/flutter/flutter/assets/59215665/e1738bd2-98d2-491b-9a4a-d2c7cbc5a080"> | <img width="25" alt="image" src="https://github.com/flutter/flutter/assets/59215665/c245d786-19aa-4a14-8df3-029591d1debd"> |

Partially addresses https://github.com/flutter/flutter/issues/29483

This will need tests, which I will add once I know which tests break due to this commit.

Blockers: 
* https://github.com/flutter/flutter/issues/152026

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

davidhicks980 committed:
Improve CupertinoPopupSurface appearance

dkwingsmt commented:
cc @matthew-carroll @xster 

The difference is phenomenal!

I have the following problems:
1. How do we verify this is the closest we can get, or I mean, how do we document these magic numbers? If future readers don't know how these numbers are derived, then we can never improve them and are essentially marking them as "magic, don't touch".
2. How do we unit test them? According to my past experience on golden tests, golden tests do not apply the filters (at least the blurring is not applied).

davidhicks980 commented:
1. Yeah that's something I will document. I started with a straight saturation filter (from https://docs.rainmeter.net/tips/colormatrix-guide/), but spent a bit of time tweaking until it looked good. I'm not actually sure if we have all of the backdrop convolutions that would be needed to perfectly recreate a surface material, as it doesn't appear to just use a gaussian blur and color filter. However, this is pretty close.

```dart 
static List<double> buildLightColorFilterMatrix() {
    const double lightLumR = 0.26;
    const double lightLumG = 0.4;
    const double lightLumB = 0.17;
    const double saturation = 2.0;
    const double sr = (1 - saturation) * lightLumR;
    const double sg = (1 - saturation) * lightLumG;
    const double sb = (1 - saturation) * lightLumB;
    return <double>[
      sr + saturation, sg, sb, 0.0, 0.0, //
      sr, sg + saturation, sb, 0.0, 0.0, //
      sr, sg, sb + saturation, 0.0, 0.0, //
      0.0, 0.0, 0.0, 1.0, 0.0, //
    ];
  }

 
  static List<double> buildDarkColorFilterMatrix() {
    const double additive = 0.3;
    const double darkLumR = 0.45;
    const double darkLumG = 0.8;
    const double darkLumB = 0.16;
    const double saturation = 1.7;
    const double sr = (1 - saturation) * darkLumR;
    const double sg = (1 - saturation) * darkLumG;
    const double sb = (1 - saturation) * darkLumB;
    return <double>[
      sr + saturation, sg, sb, 0.0, additive, //
      sr, sg + saturation, sb, 0.0, additive, //
      sr, sg, sb + saturation, 0.0, additive, //
      0.0, 0.0, 0.0, 1.0, 0.0, //
    ];
  }
```
3. I'm also not sure what the best way to unit test this would be. A basic test would just be looking up the BackdropFilter and printing out the image filter (e.g. ImageFilter.compose(source -> ColorFilter.matrix([1.74, -0.4, -0.17, 0.0, 0.0, -0.26, 1.6, -0.17, 0.0, 0.0, -0.26, -0.4, 1.83, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0]) -> blur(30.0, 30.0, clamp) -> result)). The color filter affects the appearance of the menu a bit more, so a golden may be more useful. I haven't done golden testing yet, so I need to look into it!

davidhicks980 committed:
Update dialog.dart

Added equation to derive color filter

davidhicks980 committed:
Add whitespace back

Add whitespace back

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -468,10 +533,23 @@ class CupertinoPopupSurface extends StatelessWidget {
+        outer: filter,
+        inner: ColorFilter.matrix(
+          CupertinoTheme.maybeBrightnessOf(context) == Brightness.dark
+              ? darkMatrix
+              : lightMatrix,
```
```suggestion
              ? _darkMatrix
              : _lightMatrix,
```

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -459,6 +460,70 @@ class CupertinoPopupSurface extends StatelessWidget {
+    -0.26, -0.40,  1.83, 0.00, 0.00,
+     0.00,  0.00,  0.00, 1.00, 0.00
+  ];
+  
```
```suggestion

```

dkwingsmt replied:
Extra EOL space

dkwingsmt commented:
About unit test, my idea is:
1. Blurring is hard to test, and is technically not the target of this PR. So we're only testing the change of color through this filter.
2. With these matrices, I assume we can derive what a certain color will be converted to. If this is true...
3. We can set up a few tests, each testing what a certain background color (`A`) will be converted to (`B`) by these matrices.
4. The expected color (`C`) can be obtained by drawing a dialog over a solid background of `A`.
5. `B` and `C` will not match exactly, but we can note the true value `C` in the doc, and expect the distance between `B` and `C` to be less than a certain threshold.
6. For example, we pick 64 colors by having R, G, B each being 0, 64, 128, 255, and calculate the average squared distance.

davidhicks980 committed:
Update packages/flutter/lib/src/cupertino/dialog.dart

Co-authored-by: Tong Mu <dkwingsmt@users.noreply.github.com>

davidhicks980 committed:
Update packages/flutter/lib/src/cupertino/dialog.dart

Co-authored-by: Tong Mu <dkwingsmt@users.noreply.github.com>

davidhicks980 committed:
Merge branch 'flutter:master' into patch-10

davidhicks980 committed:
Added CupertinoPopupSurface Goldens

davidhicks980 committed:
Removed whitespace

davidhicks980 committed:
Added newline between tests

davidhicks980 commented:
> About unit test, my idea is:
> 
> 1. Blurring is hard to test, and is technically not the target of this PR. So we're only testing the change of color through this filter.
> 2. With these matrices, I assume we can derive what a certain color will be converted to without going through the engine. __If this is true...__

I'm not sure if this is easily done without reaching into the engine, but correct me if I'm wrong. ColorFilters appear to be composed and applied directly by skia (at least, as far as I can tell... I'm not too familiar with the native layer of Flutter): https://github.com/flutter/flutter/blob/0a99e1f9a96d4462d634bca17ac9c16090a82bc2/packages/flutter/lib/src/rendering/layer.dart#L1827

https://api.skia.org/classSkColorFilters.html

I defined a few Goldens to test the surface changes, but I'm interested to hear if there is a better way to handle this.

davidhicks980 committed:
Reduce test set for goldens.

davidhicks980 committed:
ColorFilter is still not supported on html backend, so reverting to kIsWeb conditional.

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,124 @@
+          fit: StackFit.expand,
+          children: <Widget>[
+            for (int i = 0; i < 18; i++)
+              Positioned(
```
Can you explain what this designed is made for? I can see that it is a 3-layer stack, and the layer above and below the pop up surface are interlaced blocks of various colors, but why do we need it?

As far as I understand, we only need the pop up surface on top of a background that consists of various colors.

davidhicks980 replied:
I originally wrote it for dividers overlayed on a backdrop, so you're right, it's unnecessary. Should be fixed!

dkwingsmt commented:
I think the tests make sense, but here is what it looks like:

Unpainted:
![image](https://github.com/user-attachments/assets/9eb01083-90f0-4fb9-8ae9-b9d426ae7207)

Light: 
![image](https://github.com/user-attachments/assets/aa0f97af-6fe1-43a0-837e-bc7f5dff0e87)

I'm not sure if the light mode result has the color filter applied or it is just the translucent backdrop, but the unpainted one definitely isn't affected by anything at all. :/

dkwingsmt commented:
Ok, I've found the reasons.

First, in "unpainted" tests, the two golden files are using the same name and the doc says that the theme shouldn't affect the result. This is incorrect, because the theme does affect the result, since the color filters are different. 

Second, in "unpainted" tests, the filters isn't applied at all, because there's nothing within the filter and the layer is dropped. By adding an empty `Container` to the `child`, the result is obviously different:

![image](https://github.com/user-attachments/assets/f740a3ef-6878-4098-a067-ab7639e56c41)

While this shouldn't be a problem for production code (since there'll always be a child), I suggest making `PopUpSurface.child` `required` and explain the reason, and add an `Container` to the child in the test.

```dart
// popup_surface_test.dart, L100
            Padding(
              padding: const EdgeInsets.all(16.0),
              child: CupertinoPopupSurface(
                isSurfacePainted: isSurfacePainted,
                child: child,
              ),
            ),
```

Third, as I pointed in the comment, I wonder why a 3rd layer of 18 strips is painted on top of the pop up surface. Is it used as the master for comparison? I don't think this is very useful though. Without it the result is very obvious and beautiful:

![image](https://github.com/user-attachments/assets/1a7c663c-4766-4fa4-b8a7-5d717278bd04)

My suggestion: I thought the backdrop filters don't apply to golden images, but it turns out they do. In this case, I still think most of the test cases should be used to test the color change (since that's the most complicated part), and these test should be done without the blurring effect for accuracy, while some separate tests should be used to verify the blurring effect.

* `CupertinoPopUpSurface` should allow customizing the blurring radius. This is only used for unit tests as below.
* The current unit tests that verify the color change for both themes are great, but they should use a blurring radius of 0, should not have the "top layer" of strips, and there can be a larger variety of colors.
* I suggest we remove the unpainted tests, because in practice the pop up surface is only used with the correct background. If the surface is unpainted, then it's the children's responsibility to paint the correct background afterwards, otherwise the behavior is not guaranteed.
* One or a few additional tests should be used to verify the blurring effect. We'll probably want it to be just black and white to avoid the effect of colors.

davidhicks980 committed:
Rewrote to allow all pieces to be disabled or enabled individually.

davidhicks980 committed:
Remove parameter list from light saturation matrix

davidhicks980 committed:
Swapped the order of parameters

davidhicks980 committed:
Fixed incorrect docref

davidhicks980 committed:
Made dark mode surface test render on white background

davidhicks980 committed:
Made test image prettier and explained proxy box

davidhicks980 committed:
Removed blur 15 test.

davidhicks980 commented:
Thanks for catching that! I tend to work on this stuff at night so my minds not always sharp. I implemented your suggestions, with a few small changes. 

> CupertinoPopUpSurface should allow customizing the blurring radius. This is only used for unit tests as below.
The current unit tests that verify the color change for both themes are great, but they should use a blurring radius of 0, should not have the "top layer" of strips, and there can be a larger variety of colors.

Done.

> I suggest we remove the unpainted tests, because in practice the pop up surface is only used with the correct background. If the surface is unpainted, then it's the children's responsibility to paint the correct background afterwards, otherwise the behavior is not guaranteed.

I removed the unpainted tests where ColorFilter/Blur is displayed with isSurfacePainted == false, and added a isVibrancePainted property that lets users disable the ColorFilter. This allows us to test each individual component (ColorFilter, Blur, Surface Color) independently. There are some cases where having the ColorFilter disabled is useful, particularly if multiple CupertinoPopupSurfaces are layered or if behavior has to remain consistent with web.
 
> One or a few additional tests should be used to verify the blurring effect. We'll probably want it to be just black and white to avoid the effect of colors.

I added blur tests but used the colorful background while disabling the isVibrancePainted && isSurfacePainted parameters (e.g. the blur is now applied independently).

Here are the new images (sorry for not posting these for the last push):

Color filter
| dark | light | removed |
| - | - | - |
| ![cupertinoPopupSurface color-filter dark](https://github.com/user-attachments/assets/03d6eacc-f8d6-4a18-836a-a6305276af46) | ![cupertinoPopupSurface color-filter light](https://github.com/user-attachments/assets/98a9f52d-3673-441f-b6be-a53b7e606d8c) | ![cupertinoPopupSurface color-filter removed](https://github.com/user-attachments/assets/0c681f79-40ef-4b24-8a0e-e0a9a767d2d2) | 

Blur
| 0 | 5 | 30 (default)|
| - | - | - | 
| ![cupertinoPopupSurface blur 0](https://github.com/user-attachments/assets/99060eb4-30d7-470c-8ad4-409b1da67f50)| ![cupertinoPopupSurface blur 5](https://github.com/user-attachments/assets/23c346d6-d893-4d85-abaa-cdf63b9dcb32)| ![cupertinoPopupSurface blur 30](https://github.com/user-attachments/assets/37452133-37f5-4545-9ce6-9aca17c7b8dc)| 

Surface color
| dark | light | removed |
| - | - | - |
| ![cupertinoPopupSurface surface-color dark](https://github.com/user-attachments/assets/06e5b862-a763-46c2-b97e-8903f11719eb) | ![cupertinoPopupSurface surface-color light](https://github.com/user-attachments/assets/c29862e1-a3e8-41a9-8ea6-a6ebf8b5d8b7) | ![cupertinoPopupSurface surface-color removed](https://github.com/user-attachments/assets/3a1846cf-4a14-4065-8d48-ae94aeeea04c) |

Composition (Check things are stacked in the right order -- no dark mode test since it's the stacking that matters)
![cupertinoPopupSurface composition](https://github.com/user-attachments/assets/8744ba6f-25bb-4888-9829-3bfa248482c6)

davidhicks980 committed:
Merge branch 'master' into patch-10

davidhicks980 committed:
Toggle instead of togge

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -423,20 +424,30 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// or removed by setting the [blurSigma] to 0.
+///
+/// The saturation effect can be removed by setting [isVibrancePainted] to
+/// false. Removing the vibrance can prevent oversaturating areas where multiple
+/// [CupertinoPopupSurface]s are stacked upon eachother.
```
I suppose this is added actually for the only purpose of testing, right?

davidhicks980 replied:
Stacked colorfilters did cause some unpleasant oversaturation on the nested CupertinoMenuAnchor I was working on, to the point where I actually disabled color filters on nested layers. But to your point, this is primarily for testing convenience. Otherwise, it's could be helpful to users who want consistent appearance across web and mobile. Should I just remove the explanation or call it "debugIsVibranceEnabled"?

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -462,26 +493,148 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// [CupertinoPopupSurface].
+  ///
+  /// Setting this to 0 will remove the blur filter.
+  static const double defaultBlurSigma = 30.0;
```
While I appreciate the design of making default values a variable, the duplicate `30.0` is concerning. Maybe:
```suggestion
  static const double defaultBlurSigma = _kBlurAmount;
```

davidhicks980 replied:
Done, although I should note that _CupertinoActionSheetState used _kBlurAmount, which I replaced with CupertinoPopupSurface.defaultBlurSigma.

dkwingsmt replied:
Oh that's interesting! I wonder why it doesn't use `CupertinoPopupSurface`. It probably should.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -450,10 +461,30 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// The strength of the gaussian blur applied to the area beneath this
+  /// surface.
+  ///
+  /// Defaults to 30.0.
```
```suggestion
  /// Defaults to [defaultBlurSigma].
```

davidhicks980 replied:
Done!

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -462,26 +493,148 @@ class CupertinoPopupSurface extends StatelessWidget {
+
+  /// The default strength of the blur applied to widgets underlying a
+  /// [CupertinoPopupSurface].
+  ///
+  /// Setting this to 0 will remove the blur filter.
```
These two lines don't apply to this variable (a static value), but the field. Move them to `blurSigma`.

davidhicks980 replied:
Done! Mistake on my end.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -462,26 +493,148 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// The default corner radius of a [CupertinoPopupSurface].
+  static const BorderRadius _clipper = BorderRadius.all(Radius.circular(_kCornerRadius));
+
+  /// The [ColorFilter.matrix] used to saturate widgets underlying a
```
Can you make all docs for private variables use `//` instead of `///`? Usually it's not a big deal, but the explanations on how the matrices are derived really should use `//`.

davidhicks980 replied:
Done! I removed the interpolation brackets and code backticks as well.

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+        matchesGoldenFile('cupertinoPopupSurface.color-filter.light.png'),
+      );
+    });
+    testWidgets('Brightness.dark color filter', (WidgetTester tester) async {
```
nit: Can you add an empty line between tests?

davidhicks980 replied:
Done! Definitely looks better.

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+        ),
+      );
+
+      // Golden displays the color filter effect of the CupertinoPopupSurface
+      // when the ambient brightness is light.
```
I think it's better to move these doc to the beginning of the test to explain the enter test.

davidhicks980 replied:
Done!

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+      );
+    });
+  });
+  group('Surface Color', () {
```
A long time ago I was told that `group` is discouraged due to having some subtle problems. Although I don't think they apply to your tests specifically, I think the convention in general is to avoid using `group`s, and just putting related tests together is enough.

davidhicks980 replied:
Removed all groups

dkwingsmt replied:
I don't know... I can't speak of any concrete cases to convincingly ban it, and I believe `group`s are still used somewhat throughout the repo. I just in general avoided it because I was told and it was not hard to avoid it either. :/ Maybe we can ask in the flutter-hackers chat to see how people feel about it. And feel free to keep your groups too.

dkwingsmt replied:
And also I'm still seeing groups all around? Did you forget to push your changes?

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+      );
+    });
+    testWidgets('Setting isSurfacePainted to false removes the surface color', (WidgetTester tester) async {
+      await tester.pumpWidget(const CupertinoApp(
+        home: ColoredBox(
```
```suggestion
      await tester.pumpWidget(
        const CupertinoApp(
          home: ColoredBox(
```

And subsequent intentations.

davidhicks980 replied:
Done!

davidhicks980 replied:
Done! I should note that this was the only one I saw, so let me know if you were thinking of any other place to indent.

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+
+  // Because ImageFilter.compose is used to apply multiple filters, the order
+  // of the filters matters. As such, test that the surface effects are stacked
+  // in the correct order.
```
Can you explain what this test is for? Also, since composition order matters, can you explain this in the doc? For example, if I want to apply each effect of the pop up surface separately in two `CupertinoPopupSurface`s, what order should I use?

davidhicks980 replied:
I added a clearer doc to tests, while also including how the surface is composed in the CupertinoPopupSurface doc (I assumed this was what you meant by doc, but let me know if I'm wrong). I wasn't sure if you meant I should explicitly say

```dart
/// You can replicate the default appearance of CupertinoPopupSurface with two surfaces: 
Stack(children: <Widget>[
          CupertinoPopupSurface(
            isSurfacePainted: false,
            blurSigma: 0,
            child: SizedBox(),
          ),
          CupertinoPopupSurface(
            isVibrancePainted: false,
            child: SizedBox(),
          )
        ])
```

or if I should just explain composition order, so I kept it simple. Let me know if I should change this.

dkwingsmt commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,328 @@
+        matchesGoldenFile('cupertinoPopupSurface.blur.30.png'),
+      );
+    });
+    testWidgets('Nonpositive blurSigma removes blur', (WidgetTester tester) async {
```
Maybe we just assert in the constructor that `blurSigma >= 0`, since negative sigma doesn't make sense anyway.

davidhicks980 replied:
Done!

davidhicks980 replied:
Done! Because I had been doing simulation animations, sometimes the animation would dip into negative territory (e.g. -0.00012), which caused problems where asserts were present. However, this is an edge case and ClampedSimulation also exists, so I think your suggestion is the better way of handling things.

dkwingsmt commented:
The tests look great! I have a few comments but all of them are small.

davidhicks980 committed:
Apply fixes from review

davidhicks980 committed:
Fix whitespace

davidhicks980 committed:
Added comment regarding layering surfaces

davidhicks980 committed:
Merge branch 'master' into patch-10

davidhicks980 committed:
Merge branch 'flutter:master' into patch-10

davidhicks980 committed:
Move composition explanation

davidhicks980 commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -211,6 +211,14 @@ bool _isInAccessibilityMode(BuildContext context) {
+///
+/// For CupertinoPopupSurface, the pixels underlying the surface are first
+/// saturated with a ColorFilter, and the resulting saturated pixels are blurred
+/// with ImageFilter.blur.
```
Whoops, pushed changes!! Thanks 😅

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,38 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// The blur strength can be changed by setting [blurSigma] to a positive value,
+/// or removed by setting the [blurSigma] to 0.
+///
+/// The saturation effect can be removed by setting [isVibrancePainted] to
+/// false. Removing the vibrance can prevent oversaturating areas where multiple
+/// [CupertinoPopupSurface]s are stacked upon eachother.
```
That's a really good point! I suggest adding this point to the comment:

```suggestion
/// The saturation effect can be removed by setting [isVibrancePainted] to
/// false. The saturation effect is not supported on Web and will not be 
/// applied regardless of the [isVibrancePainted]. Removing the vibrance
/// can be helpful to achieve consistent appearance across web
/// and mobile, or to prevent oversaturating areas where multiple
/// [CupertinoPopupSurface]s are stacked upon each other.
```

davidhicks980 replied:
Done, except I changed `regardless of the [isVibrancePainted]` to `regardless of the value of [isVibrancePainted]`

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,38 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// false. Removing the vibrance can prevent oversaturating areas where multiple
+/// [CupertinoPopupSurface]s are stacked upon eachother.
+///
+/// For more complicated layouts, the surface color can be disabled by setting
+/// [isSurfacePainted] to false. This can be used to render divider gaps, such
+/// as with [CupertinoAlertDialog], or for rendering custom surface colors.
```
```suggestion
/// The surface color can be disabled by setting [isSurfacePainted] to false, which
/// is useful for more complicated layouts, such as rendering divider gaps in
/// [CupertinoAlertDialog] or rendering custom surface colors.
```

davidhicks980 replied:
Done!

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,38 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// atop a [BackdropFilter]. The surface color is light gray when the ambient
+/// [CupertinoTheme] brightness is [Brightness.light], and dark gray when the
+/// ambient brightness is [Brightness.dark]. The [BackdropFilter] applied
+/// beneath the surface saturates and blurs underlying content.
```
I suggest replacing the last two paragraphs with this.
```suggestion
/// beneath the surface saturates and blurs underlying content.
/// In other words, the overall effect is result = add_surface_color(blur(saturate(source))).
```

davidhicks980 replied:
Done! Reads a lot better now.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,38 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// [isSurfacePainted] to false. This can be used to render divider gaps, such
+/// as with [CupertinoAlertDialog], or for rendering custom surface colors.
+///
+/// CupertinoPopupSurface uses ImageFilter.compose, which applies an inner
+/// filter first, followed by an outer filter (e.g. result =
+/// outer(inner(source))).
+///
+/// For CupertinoPopupSurface, the pixels underlying the surface are first
+/// saturated with a ColorFilter, and the resulting saturated pixels are blurred
+/// with ImageFilter.blur.
```
As suggested above, the two paragraphs can be moved up and reduced to one sentence.

davidhicks980 replied:
Added

dkwingsmt reviewed (approved):
Generally looks good to me. I have some suggestion on comments.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,38 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
-/// rendering divider gaps for a more complicated layout, e.g., [CupertinoAlertDialog].
-/// Additionally, the white paint can be disabled to render a blurred rounded
-/// rectangle without any color (similar to iOS's volume control popup).
+/// By default, this surface is composed of a transparent surface color painted
```
Is "translucent" the more accurate word? IMO "transparent" means not being able to see at all.
```suggestion
/// By default, this surface is composed of a translucent surface color painted
```

davidhicks980 replied:
This is interesting and I did a little reading. Apparently translucent materials scatter light (e.g. like a blur convolution), whereas transparent materials do not (good demonstration: https://www.youtube.com/watch?v=jCnrne72W3U). Since the color is being painted on top of a blurred surface, but is not actually contributing to any pixel convolution, transparent would follow the strict definition. But to your point, "transparency" doesn't feel as right, so I'm on the wall on this one. Let me know what you think.

dkwingsmt replied:
I've also read some web pages that it seems people are not very consistent on it. Another problem is that, although the color alone is scatter-less, the overall effect of the pop up surface is with scatter. But I think using "transparent" here shouldn't cause much confusion anyway.

davidhicks980 committed:
Merge branch 'flutter:master' into patch-10

davidhicks980 committed:
Review fixes

davidhicks980 committed:
Merge branch 'master' into patch-10

davidhicks980 commented on packages/flutter/test/cupertino/popup_surface_test.dart:
```diff
@@ -0,0 +1,345 @@
+}
+
+void main() {
+    // Golden displays the color filter effect of the CupertinoPopupSurface
```
Done!

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -519,9 +533,30 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// blur filter.
+  final double blurSigma;
+
+  /// Whether or not a [ColorFilter] should be applied beneath this surface.
```
nit: specify the natural / purpose of the ColorFilter? It's only for adjusting the saturation I think?

davidhicks980 replied:
Added

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -531,26 +566,142 @@ class CupertinoPopupSurface extends StatelessWidget {
   /// The widget below this widget in the tree.
-  final Widget? child;
+  ///
+  /// Because [CupertinoPopupSurface] is composed of proxy boxes, which mimic
```
nit: I think this could just be regular code comments, since this is justification for making this field non-nullable?

davidhicks980 replied:
Fixed!

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -531,26 +566,142 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// The default corner radius of a [CupertinoPopupSurface].
+  static const BorderRadius _clipper = BorderRadius.all(Radius.circular(14));
+
+  // The ColorFilter matrix used to saturate widgets underlying a
```
uber nit: consider putting symbols such as `ColorFilter` in square brackets, they won't show up in public API docs, but the analyzer will analyze / link them.

davidhicks980 replied:
Done! Good to know.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +492,34 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+/// or removed by setting the [blurSigma] to 0.
+///
+/// The saturation effect can be removed by setting [isVibrancePainted] to
+/// false. The saturation effect is not supported on Web and will not be
```
I see it's disabled in the widget by a `kIsWeb` check. Any reason this should not be enabled on the web (instead of letting developer decide what to do)? 

Looks like the reason is the `src` blend mode? The canvaskit renderer supports that I think, so disabling this on web altogether is probably a bit of an overkill (and iirc there are plans to deprecate the HTML renderer)?

dkwingsmt replied:
I confirm that HTML renderer is getting deprecated.

davidhicks980 replied:
On the HTML renderer, I was getting this error: <img width="748" alt="image" src="https://github.com/user-attachments/assets/15354b5f-93bf-4497-bca0-428ddb681b47">

I initially thought it was the ColorFilter, but ImageFilter.compose appears to be the true culprit (I suppose it imports ImageFilter.erode? Not totally sure...)  

To your point, it does work on CanvasKit, so I revised the check to `if ((kIsWeb && !isCanvasKit) || !isVibrancePainted)`. Let me know if I should just check for !isVibrancePainted, or if this works.

dkwingsmt replied:
Have you tried the wasm implementation? I'm wondering if it's supported.

To enable it, use `--wasm`. The framework checks it with `isSkwasm`.

LongCatIsLooong replied:
The erode thing seems to be a typo:

https://github.com/flutter/engine/blob/6648b3cfe0d0f497cc033bac228751fc9d897ee3/lib/web_ui/lib/src/engine/html/renderer.dart#L139-L146

This is probably supported on wasm too, so `isSkiaWeb` probably.

LongCatIsLooong replied:
> To your point, it does work on CanvasKit, so I revised the check to if ((kIsWeb && !isCanvasKit) || !isVibrancePainted). Let me know if I should just check for !isVibrancePainted, or if this works.

This SGTM (with `s/isCanvasKit/isSkiaWeb/`). It's likely that the HTML renderer will be deprecated, then people will audit `isSkiaWeb` usages and update the implementation here.

davidhicks980 replied:
Perfect, I subbed for if ((kIsWeb && !isSkiaWeb) || !debugIsVibrancePainted)

LongCatIsLooong reviewed (approved):
LGTM with nits.

davidhicks980 committed:
Fixed reviews

davidhicks980 committed:
Remove trailing newline

davidhicks980 committed:
Add check for HTML renderer

davidhicks980 commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -79,7 +79,7 @@ const TextStyle _kActionSheetContentStyle = TextStyle(
 
 // Generic constants shared between Dialog and ActionSheet.
-const double _kBlurAmount = 20.0;
+// Blur kernel eyeballed from iOS simulator
```
Whoops, fixed! Nice cat btw.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +491,33 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
-/// rendering divider gaps for a more complicated layout, e.g., [CupertinoAlertDialog].
-/// Additionally, the white paint can be disabled to render a blurred rounded
-/// rectangle without any color (similar to iOS's volume control popup).
+/// By default, this surface is composed of a transparent surface color painted
+/// atop a [BackdropFilter]. The surface color is light gray when the ambient
+/// [CupertinoTheme] brightness is [Brightness.light], and dark gray when the
+/// ambient brightness is [Brightness.dark]. The [BackdropFilter] applied
+/// beneath the surface saturates and blurs underlying content. In other words,
+/// the overall effect is result = add_surface_color(blur(saturate(source))).
```
Sorry to bikeshed again, but I'm thinking of explaining it without technical terms, such as `BackdropFilter`.

```suggestion
/// An iOS-style component for creating modal overlays like dialogs and action sheets.
///
/// By default, [CupertinoPopupSurface] generates a rounded rectangle surface that applies two effects
/// to the background content:
///
///   1. Background filter: Increases saturation and then adds blur to the content behind
///      the surface.
///   2. Overlay color: Covers the filtered background with a transparent surface color.
///      The color adapts to the CupertinoTheme's brightness: Light gray when 
///      [CupertinoTheme] brightness is [Brightness.light], and dark gray when [Brightness.dark].
```

davidhicks980 replied:
No worries -- it does read better. The only thing I changed was 
`Increases saturation and then adds blur to the content behind the surface` =>  `Saturates and then blurs content behind the surface`, but let me know if you want me to change it back.

dkwingsmt replied:
That's definitely better!

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -519,9 +531,31 @@ class CupertinoPopupSurface extends StatelessWidget {
+  /// saturated.
+  ///
+  /// Defaults to true.
+  final bool isVibrancePainted;
```
If it turns out only HTML renderer is not supported, we might want to turn `isVibrancePainted` into `debugIsVibrancePainted` because I can't see it being very useful for production purposes for now.

davidhicks980 replied:
Changed!

davidhicks980 committed:
Review fixes.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -531,26 +562,143 @@ class CupertinoPopupSurface extends StatelessWidget {
+  ];
+
+  ImageFilter? _buildFilter(Brightness? brightness) {
+    if ((kIsWeb && !isSkiaWeb) || !debugIsVibrancePainted) {
```
Wrap `debugIsVibrancePainted` in an assert so that it only applies in debug mode.
```suggestion
    bool isVibrancePainted = true;
    assert(() {
      isVibrancePainted = debugIsVibrancePainted;
      return true;
    }());
    if ((kIsWeb && !isSkiaWeb) || !isVibrancePainted) {
```

davidhicks980 replied:
Added

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -519,9 +532,27 @@ class CupertinoPopupSurface extends StatelessWidget {
+  ///
+  /// The appearance of the [ColorFilter] is determined by the [Brightness]
+  /// value obtained from the ambient [CupertinoTheme].
+  ///
```
```suggestion
  ///
  /// The vibrance is always painted if asserts are disabled.
  ///
```

davidhicks980 replied:
Added

dkwingsmt commented:
`cupertino.cupertinoPopupSurface.surface-color.removed` turns out to be complete black in golden. I suppose it's not expected?

![image](https://github.com/user-attachments/assets/70381e0c-3406-4767-b51f-4f395741694c)

davidhicks980 committed:
Set surface background tests to colorful tiles.

davidhicks980 commented:
@dkwingsmt This was actually intentional, since the purpose of the tests was just to see whether the surface was painted with Brightness.light, Brightness.dark, and isSurfacePainted == false. That said, I went ahead and used the same colorful tile background for all tests, so all tests now have the same underlying color tiles.

davidhicks980 committed:
Merge branch 'master' into patch-10

dkwingsmt commented:
Oh, if it's expected, then we don't need to change it. I was just checking the golden file and verifying.

davidhicks980 committed:
Add debugIsVibrancePainted assert

davidhicks980 commented:
> Oh, if it's expected, then we don't need to change it. I was just checking the golden file and verifying.

Oh no worries -- it's probably better to have a consistent background between tests anyways. Also, sorry, I overlooked your comments above and forgot to push new commits. All other changes should now be implemented and pushed!

davidhicks980 committed:
Merge branch 'master' into patch-10

dkwingsmt commented:
I'm reviewing the golden files but a number of them seems to produce unexpected result. I've approved all the ones that produce expected results. Can you take a look? (Click the link in the first line of the bot's comment.)

davidhicks980 committed:
Added StackFit.expand to composition image

davidhicks980 committed:
Merge branch 'master' into patch-10

dkwingsmt commented:
The composition ones are fixed! (I've approved them.) 

All of the remaining ones seem to be on `linux-browser`, many of which are on `skwasm` and the remaining unspecified. I'm not sure but have you verified that the filter works on skwasm?

davidhicks980 commented:
It turns out I wasn't running my local Skia tests with --web-renderer=skwasm. It appears there may be a bug with Skwasm that leads to backdrop filters only applying when a decoration is applied to the child of BackdropFilter? I'm not sure why this is, but here is a test case: 

```dart
class _FilterTest extends StatelessWidget {
  const _FilterTest(Widget child, {this.brightness = Brightness.light})
      : _child = child;
  final Brightness brightness;
  final Widget _child;

  @override
  Widget build(BuildContext context) {
    final Size size = MediaQuery.sizeOf(context);
    final double tileHeight = size.height / 4;
    final double tileWidth = size.width / 8;
    return CupertinoApp(
      home: Stack(fit: StackFit.expand, children: <Widget>[
        // 512 color tiles
        // 4 alpha levels (0.416, 0.25, 0.5, 0.75)
        for (int a = 0; a < 4; a++)
          for (int h = 0; h < 8; h++) // 8 hues
            for (int s = 0; s < 4; s++) // 4 saturation levels
              for (int b = 0; b < 4; b++) // 4 brightness levels
                Positioned(
                  left: h * tileWidth + b * tileWidth / 4,
                  top: a * tileHeight + s * tileHeight / 4,
                  height: tileHeight,
                  width: tileWidth,
                  child: ColoredBox(
                    color: HSVColor.fromAHSV(
                      0.5 + a / 8,
                      h * 45,
                      0.5 + s / 8,
                      0.5 + b / 8,
                    ).toColor(),
                  ),
                ),
        Padding(
          padding: const EdgeInsets.all(32),
          child: CupertinoTheme(
            data: CupertinoThemeData(brightness: brightness),
            child: _child,
          ),
        ),
      ]),
    );
  }
}

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return _FilterTest(Row(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        DecoratedBox(
          decoration: BoxDecoration(
            border: Border.all(
                color: Colors.red,
                width: 2,
                strokeAlign: BorderSide.strokeAlignOutside),
          ),
          child: ClipRect(
            child: BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 10, sigmaY: 10),
              // ignore: sized_box_for_whitespace
              child: Container(
                height: 200,
                width: 200,
              ),
            ),
          ),
        ),
        const SizedBox(width: 8),
        DecoratedBox(
          decoration: BoxDecoration(
            border: Border.all(
              color: Colors.blue,
              width: 2,
              strokeAlign: BorderSide.strokeAlignOutside,
            ),
          ),
          child: ClipRect(
            child: BackdropFilter(
              filter: ImageFilter.blur(sigmaX: 10, sigmaY: 10),
              child: Container(
                color: Colors.transparent,
                height: 200,
                width: 200,
              ),
            ),
          ),
        ),
      ],
    ));
  }
}
```

I'll file an issue.

davidhicks980 commented:
Relevant issue https://github.com/flutter/flutter/issues/152026

Piinks commented:
Hey @davidhicks980 I have checked in on the blocking issue. WDYT of skipping these test on WASM so that we can unblock this issue? We've done this before on some web tests, and when that issue is resolved we can enable the tests. Let me know what you think. :)

davidhicks980 commented:
@Piinks That works for me! I'll update the PR tonight.

davidhicks980 committed:
Skip wasm tests that involve a backdrop filter.

davidhicks980 committed:
Merge branch 'master' into patch-10

davidhicks980 committed:
Added assertion error and clarified surface test wording.

davidhicks980 commented:
This should now be updated with skip annotations, i.e.
```dart 
skip: kIsWasm, // https://github.com/flutter/flutter/issues/152026
```

I also updated a few tests to make them more specific based on the testing guidelines -- I think I had skipped one of the testing documents from the contributing guidelines, so sorry for the late update! Largest difference was splitting 'Nonpositive blurSigma removes blur' into 'Setting blurSigma to a negative number throws' and 'Setting blurSigma to zero removes blur'.

dkwingsmt commented on packages/flutter/lib/src/cupertino/dialog.dart:
```diff
@@ -492,20 +491,34 @@ class _CupertinoAlertDialogState extends State<CupertinoAlertDialog> {
+///
+/// The saturation effect can be removed for debugging by setting
+/// [debugIsVibrancePainted] to false. The saturation effect is not supported on
+/// the html web renderer and will not be applied regardless of the value of
```
```suggestion
/// web with the skwasm renderer and will not be applied regardless of the value of
```

The HTML renderer has been deprecated and removed https://docs.flutter.dev/platform-integration/web/renderers so we only mention skwasm here.

dkwingsmt commented:
Thank you for your update! Being unsupported by skwasm is unfortunate but it'll probably be solved in the future and doesn't reduce the value of this PR. There's an update to the doc and we're ready to go.

davidhicks980 committed:
Replaced web renderer => skwasm renderer due to html web renderer being deprecated.

dkwingsmt commented:
Look perfect :)

dkwingsmt commented:
I've created an issue to remind us the skwasm issue. https://github.com/flutter/flutter/issues/154698
